### PR TITLE
bintools-wrapper: skip dynamic linker for static binaries

### PR DESCRIPTION
**Pull request by Mic92**
###### Motivation for this change

Revised version of https://github.com/NixOS/nixpkgs/pull/106908

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/bintools-wrapper/ld-wrapper.sh [hunk 1]
```diff
@@ -63,6 +69,11 @@ extraBefore=(${hardeningLDFlags[@]+"${hardeningLDFlags[@]}"})
 if [ -z "${NIX_LDFLAGS_SET_@suffixSalt@:-}" ]; then
     extraAfter+=($NIX_LDFLAGS_@suffixSalt@)
     extraBefore+=($NIX_LDFLAGS_BEFORE_@suffixSalt@)
+    # By adding dynamic linker to extraBefore we allow the users set their
+    # own dynamic linker as NIX_LD_FLAGS will override earlier set flags
+    if [[ "$setDynamicLinker" = 1 && -n "$NIX_DYNAMIC_LINKER_@suffixSalt@" ]]; then
```
Comment (Mic92) [inline pkgs/build-support/bintools-wrapper/ld-wrapper.sh:74]: This now checks if a linker is present for the selected cc-wrapper. This was not the case before and broke darwin.

**Comment by Mic92** [comment]
@dasJ could you build `nix-build -A tests.cc-wrapper -A tests.cc-wrapper-clang` on macos from this commit?

File: pkgs/build-support/bintools-wrapper/add-flags.sh [hunk 2]
```diff
@@ -25,6 +26,10 @@ if [ -e @out@/nix-support/libc-ldflags ]; then
     NIX_LDFLAGS_@suffixSalt@+=" $(< @out@/nix-support/libc-ldflags)"
 fi
 
+if [ -z "$NIX_DYNAMIC_LINKER_@suffixSalt@" ] && [ -e @out@/nix-support/dynamic-linker ]; then
```
Comment (SuperSandro2000) [inline pkgs/build-support/bintools-wrapper/add-flags.sh:29]:
```suggestion
if [ -z "$NIX_DYNAMIC_LINKER_@suffixSalt@" && -e @out@/nix-support/dynamic-linker ]; then
```

Could we do this or does that not work with sh?

**Comment by dasJ** [comment]
```
building '/nix/store/dw92gm1zc22mcn43grbpy6bdc8a84aza-Csu-osx-10.11.6.drv'...
unpacking sources
unpacking source archive /nix/store/5bdnlv9gsq6rcx22qcdq4drnlsf8vfpk-Csu-85.tar.gz
source root is Csu-85
setting SOURCE_DATE_EPOCH to timestamp 1381273907 of file Csu-85/start.s
patching sources
configuring
no configure script, doing nothing
building
build flags: -j8 -l8 SHELL=/nix/store/x9mh477dvw0929zgbpkvpxqy4ifdh5qj-bootstrap-tools/bin/bash
Makefile:153: target '/lib' given more than once in the same rule
cc  -r  -Os -mmacosx-version-min=10.6 -mdynamic-no-pic -nostdlib -keep_private_externs start.s crt.c dyld_glue.s -o crt1.v1.o  -DCRT -DOLD_LIBSYSTEM_SUPPORT
cc  -r  -Os -mmacosx-version-min=10.6 -nostdlib -keep_private_externs start.s crt.c dyld_glue.s -o crt1.v2.o  -DCRT
cc  -r  -Os -mmacosx-version-min=10.6 -nostdlib -keep_private_externs start.s crt.c -o crt1.v3.o  -DADD_PROGRAM_VARS
cc  -r  -Os -mmacosx-version-min=10.6 -nostdlib -keep_private_externs start.s crt.c -o crt1.v4.o  -DADD_PROGRAM_VARS
cc  -r  -Os -mmacosx-version-min=10.6 -nostdlib -keep_private_externs start.s crt.c dyld_glue.s -o gcrt1.o  -DGCRT  -DOLD_LIBSYSTEM_SUPPORT
cc  -r  -Os -static -Wl,-new_linker -nostdlib -keep_private_externs start.s crt.c -o crt0.o
cc  -r  -Os -mmacosx-version-min=10.6  -nostdlib -keep_private_externs dyld_glue.s icplusplus.c -o dylib1.v1.o  -DCFM_GLUE
cc  -r  -Os -mmacosx-version-min=10.6  -nostdlib -keep_private_externs dyld_glue.s -o dylib1.v2.o  -DCFM_GLUE
ld: unknown option: -dynamic-linker=/usr/lib/dyld
ld: unknown option: -dynamic-linker=/usr/lib/dyld
ld: unknown option: -dynamic-linker=/usr/lib/dyld
ld: unknown option: -dynamic-linker=/usr/lib/dyld
ld: unknown option: -dynamic-linker=/usr/lib/dyld
ld: unknown option: -dynamic-linker=/usr/lib/dyld
ld: unknown option: -dynamic-linker=/usr/lib/dyld
clang-7clang-7clang-7: : clang-7: : clang-7errorerror: errorerror: : : : error: linker command failed with exit code 1 (use -v to see invocation)
linker command failed with exit code 1 (use -v to see invocation)linker command failed with exit code 1 (use -v to see invocation)linker command failed with exit code 1 (use -v to see invocation)

linker command failed with exit code 1 (use -v to see invocation)

clang-7clang-7: : errorerror: : linker command failed with exit code 1 (use -v to see invocation)linker command failed with exit code 1 (use -v to see invocation)

make: *** [Makefile:95: dylib1.v2.o] Error 1
make: *** Waiting for unfinished jobs....
make: *** [Makefile:81: crt1.v4.o] Error 1
make: *** [Makefile:92: dylib1.v1.o] Error 1
make: *** [Makefile:75: crt1.v2.o] Error 1
make: *** [Makefile:84: gcrt1.o] Error 1
make: *** [Makefile:72: crt1.v1.o] Error 1
make: *** [Makefile:78: crt1.v3.o] Error 1
builder for '/nix/store/dw92gm1zc22mcn43grbpy6bdc8a84aza-Csu-osx-10.11.6.drv' failed with exit code 2
building '/nix/store/i3dy62d7lkq4q7y9h1qfan3wmfripk1g-ICU-osx-10.10.5.drv'...
cannot build derivation '/nix/store/55147vsyg0yn4r9ihslq1iivnlsgwhxg-Libsystem-osx-10.12.6.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/5i0mk08mpbyxmx3w1ihbkyqkpzv2farz-binutils-wrapper-2.31.1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/72qlykisycs8k1ijbkjgdqkn3vy30v8z-bootstrap-stage0-clang-wrapper-7.1.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/7i75jzy70kna2w64mxfcn5mxmllz65qb-bootstrap-stage0-clang-wrapper-7.1.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/pz8rzm7pc5736rjm94rgwd9g4bh0jkij-bootstrap-stage2-stdenv-darwin.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/qzj31ki0xjssjx3vdwz9qq3lzf265ckw-bootstrap-stage2-stdenv-darwin.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/0ykdcjc9rs4c1v704i9zdfhy89dcqf2h-bootstrap-stage3-stdenv-darwin.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/9j9sabnb199hv8zzl2qn940r3h43ivrl-bootstrap-stage3-stdenv-darwin.drv': 1 dependencies couldn't be built
building '/nix/store/51ypnbfgkzavfxb7vy1zrm5vvwbdmss0-c-ares-1.17.1.drv'...
cannot build derivation '/nix/store/8c83x986by46g1zcnaddn618yrlk9p0a-clang-wrapper-7.1.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/if61idhrb1qyrrs6178n5pn3lhhgn03h-clang-wrapper-7.1.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/0pddc1klap22v7ry13jkk65vih3gg7k5-perl-5.32.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/5pyqwhz45k89jwk80hpywx3g9sr5xiz3-stdenv-darwin.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/2wsmcf7xdcd5wv39xvlkpd2736nd6w77-bash-4.4-p23.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/s7n5dm941pcfs1d50hjxr6kck4j4fr9d-cc-wrapper-test.drv': 1 dependencies couldn't be built
error: build of '/nix/store/bw2221k52bn5qivb7kv5wgyvq82a9zps-cc-wrapper-test.drv', '/nix/store/s7n5dm941pcfs1d50hjxr6kck4j4fr9d-cc-wrapper-test.drv' failed
[1]    11276 exit 100   nix-build -A tests.cc-wrapper -A tests.cc-wrapper-clang
```

File: pkgs/build-support/bintools-wrapper/add-flags.sh (hunk 2 above)
Comment (dasJ) [reply pkgs/build-support/bintools-wrapper/add-flags.sh:29]: That's not how shell works

Comment (SuperSandro2000) [reply pkgs/build-support/bintools-wrapper/add-flags.sh:29]: You mean sh? Because bash works like this.

**Comment by SuperSandro2000** [comment]
> ```
> building '/nix/store/dw92gm1zc22mcn43grbpy6bdc8a84aza-Csu-osx-10.11.6.drv'...
> unpacking sources
> unpacking source archive /nix/store/5bdnlv9gsq6rcx22qcdq4drnlsf8vfpk-Csu-85.tar.gz
> source root is Csu-85
> setting SOURCE_DATE_EPOCH to timestamp 1381273907 of file Csu-85/start.s
> patching sources
> configuring
> no configure script, doing nothing
> building
> build flags: -j8 -l8 SHELL=/nix/store/x9mh477dvw0929zgbpkvpxqy4ifdh5qj-bootstrap-tools/bin/bash
> Makefile:153: target '/lib' given more than once in the same rule
> cc  -r  -Os -mmacosx-version-min=10.6 -mdynamic-no-pic -nostdlib -keep_private_externs start.s crt.c dyld_glue.s -o crt1.v1.o  -DCRT -DOLD_LIBSYSTEM_SUPPORT
> cc  -r  -Os -mmacosx-version-min=10.6 -nostdlib -keep_private_externs start.s crt.c dyld_glue.s -o crt1.v2.o  -DCRT
> cc  -r  -Os -mmacosx-version-min=10.6 -nostdlib -keep_private_externs start.s crt.c -o crt1.v3.o  -DADD_PROGRAM_VARS
> cc  -r  -Os -mmacosx-version-min=10.6 -nostdlib -keep_private_externs start.s crt.c -o crt1.v4.o  -DADD_PROGRAM_VARS
> cc  -r  -Os -mmacosx-version-min=10.6 -nostdlib -keep_private_externs start.s crt.c dyld_glue.s -o gcrt1.o  -DGCRT  -DOLD_LIBSYSTEM_SUPPORT
> cc  -r  -Os -static -Wl,-new_linker -nostdlib -keep_private_externs start.s crt.c -o crt0.o
> cc  -r  -Os -mmacosx-version-min=10.6  -nostdlib -keep_private_externs dyld_glue.s icplusplus.c -o dylib1.v1.o  -DCFM_GLUE
> cc  -r  -Os -mmacosx-version-min=10.6  -nostdlib -keep_private_externs dyld_glue.s -o dylib1.v2.o  -DCFM_GLUE
> ld: unknown option: -dynamic-linker=/usr/lib/dyld
> ld: unknown option: -dynamic-linker=/usr/lib/dyld
> ld: unknown option: -dynamic-linker=/usr/lib/dyld
> ld: unknown option: -dynamic-linker=/usr/lib/dyld
> ld: unknown option: -dynamic-linker=/usr/lib/dyld
> ld: unknown option: -dynamic-linker=/usr/lib/dyld
> ld: unknown option: -dynamic-linker=/usr/lib/dyld
> clang-7clang-7clang-7: : clang-7: : clang-7errorerror: errorerror: : : : error: linker command failed with exit code 1 (use -v to see invocation)
> linker command failed with exit code 1 (use -v to see invocation)linker command failed with exit code 1 (use -v to see invocation)linker command failed with exit code 1 (use -v to see invocation)
> 
> linker command failed with exit code 1 (use -v to see invocation)
> 
> clang-7clang-7: : errorerror: : linker command failed with exit code 1 (use -v to see invocation)linker command failed with exit code 1 (use -v to see invocation)
> 
> make: *** [Makefile:95: dylib1.v2.o] Error 1
> make: *** Waiting for unfinished jobs....
> make: *** [Makefile:81: crt1.v4.o] Error 1
> make: *** [Makefile:92: dylib1.v1.o] Error 1
> make: *** [Makefile:75: crt1.v2.o] Error 1
> make: *** [Makefile:84: gcrt1.o] Error 1
> make: *** [Makefile:72: crt1.v1.o] Error 1
> make: *** [Makefile:78: crt1.v3.o] Error 1
> builder for '/nix/store/dw92gm1zc22mcn43grbpy6bdc8a84aza-Csu-osx-10.11.6.drv' failed with exit code 2
> building '/nix/store/i3dy62d7lkq4q7y9h1qfan3wmfripk1g-ICU-osx-10.10.5.drv'...
> cannot build derivation '/nix/store/55147vsyg0yn4r9ihslq1iivnlsgwhxg-Libsystem-osx-10.12.6.drv': 1 dependencies couldn't be built
> cannot build derivation '/nix/store/5i0mk08mpbyxmx3w1ihbkyqkpzv2farz-binutils-wrapper-2.31.1.drv': 1 dependencies couldn't be built
> cannot build derivation '/nix/store/72qlykisycs8k1ijbkjgdqkn3vy30v8z-bootstrap-stage0-clang-wrapper-7.1.0.drv': 1 dependencies couldn't be built
> cannot build derivation '/nix/store/7i75jzy70kna2w64mxfcn5mxmllz65qb-bootstrap-stage0-clang-wrapper-7.1.0.drv': 1 dependencies couldn't be built
> cannot build derivation '/nix/store/pz8rzm7pc5736rjm94rgwd9g4bh0jkij-bootstrap-stage2-stdenv-darwin.drv': 1 dependencies couldn't be built
> cannot build derivation '/nix/store/qzj31ki0xjssjx3vdwz9qq3lzf265ckw-bootstrap-stage2-stdenv-darwin.drv': 1 dependencies couldn't be built
> cannot build derivation '/nix/store/0ykdcjc9rs4c1v704i9zdfhy89dcqf2h-bootstrap-stage3-stdenv-darwin.drv': 1 dependencies couldn't be built
> cannot build derivation '/nix/store/9j9sabnb199hv8zzl2qn940r3h43ivrl-bootstrap-stage3-stdenv-darwin.drv': 1 dependencies couldn't be built
> building '/nix/store/51ypnbfgkzavfxb7vy1zrm5vvwbdmss0-c-ares-1.17.1.drv'...
> cannot build derivation '/nix/store/8c83x986by46g1zcnaddn618yrlk9p0a-clang-wrapper-7.1.0.drv': 1 dependencies couldn't be built
> cannot build derivation '/nix/store/if61idhrb1qyrrs6178n5pn3lhhgn03h-clang-wrapper-7.1.0.drv': 1 dependencies couldn't be built
> cannot build derivation '/nix/store/0pddc1klap22v7ry13jkk65vih3gg7k5-perl-5.32.0.drv': 1 dependencies couldn't be built
> cannot build derivation '/nix/store/5pyqwhz45k89jwk80hpywx3g9sr5xiz3-stdenv-darwin.drv': 1 dependencies couldn't be built
> cannot build derivation '/nix/store/2wsmcf7xdcd5wv39xvlkpd2736nd6w77-bash-4.4-p23.drv': 1 dependencies couldn't be built
> cannot build derivation '/nix/store/s7n5dm941pcfs1d50hjxr6kck4j4fr9d-cc-wrapper-test.drv': 1 dependencies couldn't be built
> error: build of '/nix/store/bw2221k52bn5qivb7kv5wgyvq82a9zps-cc-wrapper-test.drv', '/nix/store/s7n5dm941pcfs1d50hjxr6kck4j4fr9d-cc-wrapper-test.drv' failed
> [1]    11276 exit 100   nix-build -A tests.cc-wrapper -A tests.cc-wrapper-clang
> ```

👍🏼

File: pkgs/build-support/bintools-wrapper/add-flags.sh (hunk 2 above)
Comment (dasJ) [reply pkgs/build-support/bintools-wrapper/add-flags.sh:29]: No?

```shell
$ echo $BASH_VERSION
4.4.23(1)-release
$ [ -z lal && -n lal ]
bash: [: missing `]'
```

Comment (Ericson2314) [reply pkgs/build-support/bintools-wrapper/add-flags.sh:29]: maybe needs `[[`...`]]`

Comment (SuperSandro2000) [reply pkgs/build-support/bintools-wrapper/add-flags.sh]: > maybe needs `[[`...`]]`

Yeah, that only work with new test. nvm.

**Comment by SuperSandro2000** [comment]
```
ryazs91-xnu-osx-10.12.6/include /private/var/folders/r4/k85s52jx3555ckj0wty615s00000gp/T/nix-build-xnu-osx-10.12.6.drv-0/xnu-3789.70.16/libsyscall/mach/vm_map.defs                         post-installation fixup                                                                                                                                                                     rewriting symlink /nix/store/ccbz5jm10abnl2qac6lyv7y2iryazs91-xnu-osx-10.12.6/Library/Frameworks/System.framework/Versions/Current to be relative to /nix/store/ccbz5jm10abnl2qac6lyv7y2iryazs91-xnu-osx-10.12.6                                                                                                                                                                        rewriting symlink /nix/store/ccbz5jm10abnl2qac6lyv7y2iryazs91-xnu-osx-10.12.6/Library/Frameworks/System.framework/Headers to be relative to /nix/store/ccbz5jm10abnl2qac6lyv7y2iryazs91-xnu-osx-10.12.6                                                                                                                                                                                 strip is /nix/store/x9mh477dvw0929zgbpkvpxqy4ifdh5qj-bootstrap-tools/bin/strip                                                                                                              building '/nix/store/5dds2grivw41y3x6l5qnzfssyfmy90bv-swift-corefoundation.drv'...                                                                                                          unpacking sources                                                                                                                                                                           unpacking source archive /nix/store/q19wpy2aylv4a6z5c6p52ih8pi2kcbpq-source                                                                                                                 source root is source/CoreFoundation                                                                                                                                                        patching sources                                                                                                                                                                            substituteStream(): WARNING: pattern '#if defined(__GNU__) ||' doesn't match anything in file 'Base.subproj/CFAsmMacros.h'                                                                  configuring                                                                                                                                                                                 /nix/store/jfq0n06qfdwmwgq19643z7fr1g5inbsr-bootstrap-stage1-stdenv-darwin/setup: ../configure: /usr/bin/env: bad interpreter: Operation not permitted                                      builder for '/nix/store/5dds2grivw41y3x6l5qnzfssyfmy90bv-swift-corefoundation.drv' failed with exit code 126
```

I think this is unrelated to this PR.

**Comment by Mic92** [comment]
@SuperSandro2000 what did you try to build? Can you disable the sandbox just for:

```console
$ nix-build -A tests.cc-wrapper -A tests.cc-wrapper-clang
```

**Comment by dasJ** [comment]
@Mic92 since others seem to do the build stuff, I'll unsubscribe here. Feel free to ping me if you need me again

**Comment by SuperSandro2000** [comment]
> @SuperSandro2000 what did you try to build? Can you disable the sandbox just for:
> 
> ```
> $ nix-build -A tests.cc-wrapper -A tests.cc-wrapper-clang
> ```

Works without sandbox 👍🏼

**Review by Emantor — APPROVED** [review]

**Review by matthewbauer — APPROVED** [review]

**Pull request merged**